### PR TITLE
Create a new template for the home page

### LINKED ISSUE #367: New template for the front page
# Objetivos / Objectivos / Purpose

This is the proposed template for the front page for discussion. The figure shows two possibilities for the rows: "Alabama-style" or "Alaska-style".  

image

- How difficult/easy it is to implement that?
- What do we need?

### PR DESCRIPTION
**Pull request by esloch**
- Closes #367.
- New materialized view created from tables historico_alerta's.
- Create a stack, scatter and indicator chart.
- Added AJAX for improved loading.
- Add tabs each disease.
- Add more information (latest update, total cities by state).
- Created cache for homepage dbdada queries.

### CODE REVIEW

**Comment by esloch** [comment]
##### Start charged 
```34 requests
701 kB transferred
1.7 MB resources
Finish: 974 ms
```
new_template_version_

**Comment by esloch** [comment]
##### Initial loading
```34 requests
701 kB transferred
1.7 MB resources
Finish: 974 ms
```
new_template_version_

**Comment by fccoelho** [comment]
O layout está legal. O volume de downloads não.
tente fazer uma discriminação de como estes 10MB estão divididos, ou seja o que está incluído neste pacote.
porque certamente não pode ser os dados que estão sendo apresentados, pois tipicamente um número ocupa 4 bytes (32bits) logo 10M corresponderiam a 2.5 milhões de números(aproximadamente), que é muito mais do que o que precisamos para a visualização de um estado. caluculei aproximadamente 90 números para cada estado, incluindo as três doenças.
e os 33MB de resources o que contém?

File: AlertaDengue/dados/charts/home.py [hunk 1]
```diff
@@ -3,174 +3,236 @@
...
 
 import plotly.graph_objs as go
-import pandas as pd
-from plotly.subplots import make_subplots
 
+import plotly.express as px
+
+from dados.dbdata import con_table
 
-class HomeCharts:
```
Comment (fccoelho) [inline AlertaDengue/dados/charts/home.py:14]: Completamente desnecessário agregar estas funções dentro de uma classe se você vai apenas usá-las como funções. Use funções simples. menos linhas de código.

File: AlertaDengue/dados/charts/home.py [hunk 2]
```diff
@@ -3,174 +3,236 @@
...
-
-        if series.empty:
-            return {
-                'ufs': [],
-                'start': None,
-                'series': {},
-                'series_est': {},
-                'disease': disease,
-            }
+    def prepare_data(cls, uf, disease):
```
Comment (fccoelho) [inline AlertaDengue/dados/charts/home.py:18]: Esta função deveria estar cacheada, com lru_cache ou outra solução

File: AlertaDengue/dados/charts/home.py [hunk 3]
```diff
@@ -3,174 +3,236 @@
...
-                go.Scatter(
-                    x=df_ufs.index,
-                    y=df_ufs[k],
-                    name=k,
-                    marker={'color': cls.colors.get(k)},
-                    text=df_ufs.index.strftime('%d-%b-%Y <br>{}'.format(k)),
-                    hovertemplate=_(
-                        '%{text} <br>' '%{y} Casos Estimados' '<extra></extra>'
-                    ),
+    def _create_scatter_chart(cls, uf, disease):
```
Comment (fccoelho) [inline AlertaDengue/dados/charts/home.py:32]: outra função candidata a cache

File: AlertaDengue/dados/templates/components/home/collapse.html [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+          </div>
+      </div>
+    </div>
+  </div>
+
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+  
+  <script> //As colunas dos gráficos são provistas pelo template da views na url
+
+  $(document).ready(function(){
```
Comment (fccoelho) [inline AlertaDengue/dados/templates/components/home/collapse.html:155]: Esta função é idêntica a todas abaixo, não precisa ser repetida basta uma vez e a cada "flip" recebe uma id  `uf_namex` diferente como argumento:
```javascript
.click(function(id){...}
```

File: AlertaDengue/dados/templates/components/home/collapse.html [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+          $("#panel1").html(data);
+          },
+          dataType: "html"
+      });
+    });
+    $("#flip2").click(function(){
+      var uf_name = document.getElementById('uf_name2').innerText;
+      var url = '/chartshome/' + uf_name;
+      $.ajax({
+        url: "/chartshome/Rio Grande do Sul",
```
Comment (fccoelho) [inline AlertaDengue/dados/templates/components/home/collapse.html:183]: 'Rio Grande do Sul' é `uf_name`. não deve estar hardcoded.

**Comment by esloch** [comment]
@fccoelho, @xmnlab sopló estos parámetros:
>    return fig.to_html(full_html=False, include_plotlyjs=False)
```
36 / 37 requests
26.1 kB / 26.1 kB transferred
5.2 MB / 5.2 MB resources
Finish: 1.2 min
DOMContentLoaded: 437 ms
Load: 1.13 s
```
new_load_template_

**Comment by fccoelho** [comment]
tempo de execução da view, no backend continua _muuuito_ lenta:
image

**Comment by fccoelho** [comment]
A performance do carregamento melhorou, mas ainda pode ser melhorada. Cada vez que se reabre uma mesma seção do acordeon, a requisição é feita novamente, e como a maior parte do tempo da requisição ocorre no backend, não há qualquer ganho de cache do navegador. Os dados baixados nas requisições devem ser caheados localmente, e requisição só reve ser feita no caso do dado não estar já presente. por exemplo, [como ilustrado aqui](https://stackoverflow.com/questions/44672327/prevent-duplicate-ajax-calls-using-result-cache).

**Comment by fccoelho** [comment]
@esloch  vc fez algum progresso sobre o cacheamento das funções em Dbdata? Note que não basta transformar as consultas em Materialized views pois isto não resolve o custo de conexão e transferência dos dados do banco para o Python.
Ao usar o [Memcached](https://devcenter.heroku.com/articles/django-memcache) nas funções do Python, o dado fica na memória dos aplicativo e portando a view pode responder instantaneamente.

**Comment by esloch** [comment]
> @esloch vc fez algum progresso sobre o cacheamento das funções em Dbdata? Note que não basta transformar as consultas em Materialized views pois isto não resolve o custo de conexão e transferência dos dados do banco para o Python.
> Ao usar o [Memcached](https://devcenter.heroku.com/articles/django-memcache) nas funções do Python, o dado fica na memória dos aplicativo e portando a view pode responder instantaneamente.

 All right, I'll work on it!

**Comment by fccoelho** [comment]
Os títulos dos eixos x dos gráficos da direita e esquerda devem ser "Semana epidemiológica" e os tick labels, ao invés de serem "2021/18" devem ser apenas o numero da semana: 18.

**Comment by fccoelho** [comment]
Os popups do gráfico da direita devem ser escritos da seguinte maneira:
"20 cidades (x%) em alerta amarelo na semana 21" ao invés de como está agora

**Comment by fccoelho** [comment]
Os popups do gráfico da direita devem ser escritos da seguinte maneira:
"20 cidades (x%) em alerta amarelo na semana 21" ao invés de como está agora

**Comment by fccoelho** [comment]
@esloch besides completing the translation of the new textual elements, we need to change the behavior in the accordion widget:

   In states where there is no data for Zika, for example, instead of showing the message "No available data..." just disable the Zika tab(make it gray)

**Comment by fccoelho** [comment]
@esloch I think we can consider a merge of the PR. It's good enough to enter production.
Then other minor tweaks can be done on separate issues, such as for example, update of the translations of the new components.

File: AlertaDengue/dados/templates/main.html [hunk 6]
```diff
@@ -1,224 +1,104 @@
...
 {% load i18n %}
 {% load static %}
+{% load home_components %}
 {% load lookup %}
-{% block extra_head %}
-    <meta property="og:image"
-          content="http://info.dengue.mat.br/static/img/screenshot.png"/>
-    <meta property="og:image:secure_url"
-          content="https://info.dengue.mat.br/static/img/screenshot.png"/>
-    <style>
```
Comment (fccoelho) [inline AlertaDengue/dados/templates/main.html:10]: In the future we should keep CSS in separate files, but that should be done on a new issue

**Review by fccoelho — APPROVED** [review]
